patches/api/0343-WIP-Brain-API.patch

```diff
@@ -0,0 +1,523 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Owen1212055 <23108066+Owen1212055@users.noreply.github.com>
+Date: Mon, 29 Nov 2021 21:57:09 -0500
+Subject: [PATCH] WIP Brain API
+
+
+diff --git a/src/main/java/io/papermc/paper/entity/brain/BrainHolder.java b/src/main/java/io/papermc/paper/entity/brain/BrainHolder.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..21e0f2fa10766b257cac7f21e5389255e77b7647
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/entity/brain/BrainHolder.java
+@@ -0,0 +1,83 @@
++package io.papermc.paper.entity.brain;
++
++import io.papermc.paper.entity.brain.activity.ActivityKey;
++import io.papermc.paper.entity.brain.activity.behavior.BehaviorPair;
++import io.papermc.paper.entity.brain.memory.MemoryKeyStatus;
++import io.papermc.paper.entity.brain.memory.MemoryPair;
++import io.papermc.paper.entity.brain.sensor.Sensor;
++import io.papermc.paper.entity.brain.sensor.SensorKey;
++import org.bukkit.entity.LivingEntity;
++import org.bukkit.entity.memory.MemoryKey;
++import org.jetbrains.annotations.NotNull;
++
++import javax.annotation.Nullable;
++import java.util.Collection;
++
++/*
++
++Initial implementation details:
++
++All entities have a brain, but not all entities actually tick their brain.
++
++
++Behaviors are just goals combined with eachother, and are defined by "Activities".
++
++Sensors populate memories, and memories are used to trigger/read by activities.
++
++
++ */
++public interface BrainHolder<T extends LivingEntity> {
```

Comment:
I think it would be better to expose a `Brain` object instead of an interface like this that entities extend.

Reply:
How would I handle the generic type then? Additionally, not all entities use their brain so it may be confusing to some people.

Should I make it so that this interface has a method to RETURN a brain instance? I'd love a pointer here.

Reply:
All `LivingEntity` have a `Brain` in vanilla, for what it is worth.

Reply:
yeah, but not all, in fact most make no use of it iirc. So I do kinda think its best to only add it to mobs that make use of it.

Reply:
That is true, but only some of them actually tick their brain. So if we were to expose this api to all entities, it may be optimal to force tick the brain for these types of entities? I am not sure, additionally, how would I handle the generics here?